topic/service.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package topic
+
+import (
+	"github.com/m3db/m3cluster/kv"
+)
+
+type service struct {
+	store kv.Store
+}
+
+// NewService creates a topic service.
+func NewService(sOpts ServiceOptions) (Service, error) {
+	store, err := sOpts.ConfigService().Store(sOpts.KVOverrideOptions())
+	if err != nil {
+		return nil, err
+	}
+	return &service{
+		store: store,
+	}, nil
+}
+
+func (s *service) Get(name string) (Topic, int, error) {
+	value, err := s.store.Get(key(name))
+	if err != nil {
+		return nil, 0, err
+	}
+	t, err := NewTopicFromValue(value)
+	if err != nil {
+		return nil, 0, err
+	}
+	return t, value.Version(), nil
+}
+
+func (s *service) CheckAndSet(name string, version int, t Topic) error {
+	pb, err := ToProto(t)
+	if err != nil {
+		return err
+	}
+	_, err = s.store.CheckAndSet(key(name), version, pb)
+	return err
+}
+
+func (s *service) Delete(name string) error {
+	_, err := s.store.Delete(key(name))
+	return err
+}
+
+func (s *service) Watch(name string) (Watch, error) {
+	w, err := s.store.Watch(key(name))
+	if err != nil {
+		return nil, err
+	}
+	return NewWatch(w), nil
+}
+
+func key(name string) string {
+	return name
+}
+
+// NewWatch creates a new topic watch.
+func NewWatch(watch kv.ValueWatch) Watch {
+	return &w{w: watch}
+}
+
+type w struct {
+	w kv.ValueWatch
+}
+
+func (w *w) C() <-chan struct{} {
+	return w.w.C()
+}
+
+func (w *w) Get() Topic {
+	value := w.w.Get()
+	if value == nil {
+		return nil
+	}
+	t, err := NewTopicFromValue(value)
+	if err != nil {
```

Comment:
Why do we want to ignore errors here (or rather in the Watch method on the interface)?

Reply:
Basically this happens when there is an invalid topic proto in KV and the user could probably only treat that case as if the topic was deleted from KV, so it returns a nil Topic here.

Reply:
Can we return an exported error so user can make that decision instead of the library making the decision for them?

Reply:
sure can expose error through the interface